### PR TITLE
Hole Punching

### PR DESCRIPTION
**Pull request by diegomrsantos**
This PR implements end-to-end TCP hole punching, integrating the previously implemented services (Autorelay and Autonat) and protocol (DCUtr) into the Hole Punching Service. 

One of the Hole Punching service's responsibilities is to detect new relayed connections and try a unilateral connection upgrade if the remote peer has public dialable addresses. 

It also required changing how local ports are chosen when dialing peers. Previously, it was always chosen by the OS according to availability. This is still done when peer reachability is unknown or public, but the listening port is used instead when the peer is unreachable.

### CODE REVIEW

File: libp2p/services/autonatservice.nim [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  AutonatService* = ref object of Service
+    newPeerHandler: PeerEventHandler
+    networkReachability: NetworkReachability
+    t: CountTable[NetworkReachability]
+    autonat: Autonat
+    newStatusHandler: NewStatusHandler
+    maxConfidence: int
+
+  NetworkReachability* {.pure.} = enum
+    Private, Public, Unknown
```
Comment (alrevuelta) [inline libp2p/services/autonatservice.nim:33]: Perhaps `Reachable`, `NotReachable`?

File: libp2p/multiaddress.nim [hunk 2]
```diff
@@ -1123,5 +1123,5 @@ proc getRepeatedField*(pb: ProtoBuffer, field: int,
         value.add(ma.get())
       else:
         value.setLen(0)
-        return err(ProtoError.IncorrectBlob)
+        #return err(ProtoError.IncorrectBlob)
```
Comment (Menduist) [inline libp2p/multiaddress.nim:1126]: This needs a better fix

File: libp2p/services/hpservice.nim [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  {.push raises: [].}
+
+import std/[tables, sequtils]
+import ../switch, ../wire
+import ../protocols/rendezvous
+import ../services/autorelayservice
+import ../discovery/[rendezvousinterface, discoverymngr]
+import ../protocols/connectivity/relay/relay
+import ../protocols/connectivity/autonat/service
+from ../protocols/connectivity/dcutr/core import DcutrError
```
Comment (Menduist) [inline libp2p/services/hpservice.nim:22]: Maybe dctur/client & server should export it instead?

File: libp2p/services/hpservice.nim [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    isPublicIPAddrProc: IsPublicIPAddrProc
+
+  IsPublicIPAddrProc* = proc(ta: TransportAddress): bool {.gcsafe, raises: [Defect].}
+
+proc new*(T: typedesc[HPService], autonatService: AutonatService, autoRelayService: AutoRelayService,
+          isPublicIPAddrProc: IsPublicIPAddrProc = isGlobal): T =
+  return T(autonatService: autonatService, autoRelayService: autoRelayService, isPublicIPAddrProc: isPublicIPAddrProc)
+
+proc tryStartingDirectConn(self: HPService, switch: Switch, peerId: PeerId): Future[bool] {.async.} =
+  await sleepAsync(100.milliseconds) # wait for AddressBook to be populated
```
Comment (Menduist) [inline libp2p/services/hpservice.nim:44]:
```suggestion
  await sleepAsync(500.milliseconds) # wait for AddressBook to be populated
```
To be safe

**Review by Menduist — DISMISSED** [review]
LGTM except https://github.com/status-im/nim-libp2p/pull/881

**Comment by diegomrsantos** [comment]
@Menduist the tests have been fixed.

**Review by Menduist — APPROVED** [review]
🎉

**Pull request merged**